File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.java
// 숫자 20 개
// 2**20 백만
// dfs 가능

class Solution {
    int answer = 0;
    
    public int solution(int[] numbers, int target) {
        
        dfs(numbers, 0, 0, target);
        
        return answer;
    }
    
    public void dfs(int[] numbers, int index, int value, int target) {
        if (index == numbers.length) {
            if (target == value) {
                answer++;
            }
            return;
        }
        
        dfs(numbers, index+1, value - numbers[index], target);
        dfs(numbers, index+1, value + numbers[index], target);
    }
}

File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.py
# 순서 바꾸지 않고 더하기 빼기만 해서 타겟넘버 만들기
# 방법의 수 return
# 완탐 가능한가?
# 20개
# 더하기 빼기 2**20 = 1048576 가능
# dfs

answer = 0
def solution(numbers, target):
    n = len(numbers)
    def dfs(index, s):
        if index == n:
            if s == target:
                global answer
                answer += 1
            return
        
        dfs(index+1, s + numbers[index])
    
        dfs(index+1, s + (numbers[index] * -1))
        
    dfs(0, 0)
    
    return answer